src/util/pointabilities.cpp

```diff
@@ -0,0 +1,143 @@
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "pointabilities.h"
+
+#include "serialize.h"
+#include "exceptions.h"
+#include <sstream>
+
+PointabilityType Pointabilities::deSerializePointabilityType(std::istream &is)
+{
+	PointabilityType pointable_type = static_cast<PointabilityType>(readU8(is));
+	switch(pointable_type) {
+		case PointabilityType::POINTABLE:
+		case PointabilityType::POINTABLE_NOT:
+		case PointabilityType::POINTABLE_BLOCKING:
+			break;
+		default:
+			pointable_type = PointabilityType::POINTABLE;
+			break;
+    }
+    return pointable_type;
+}
+
+void Pointabilities::serializePointabilityType(std::ostream &os, PointabilityType pointable_type)
+{
+	writeU8(os, static_cast<u8>(pointable_type));
+}
+
+std::string Pointabilities::toStringPointabilityType(PointabilityType pointable_type)
+{
+	switch(pointable_type) {
+		case PointabilityType::POINTABLE:
+			return "true";
+		case PointabilityType::POINTABLE_NOT:
+			return "false";
+		case PointabilityType::POINTABLE_BLOCKING:
+			return "\"blocking\"";
+		default:
+			return "unknown";
+    }
+}
+
+std::optional<PointabilityType> Pointabilities::matchNode(const std::string &name,
+	const ItemGroupList &groups) const
+{
+	auto i = nodes.find(name);
+	return i == nodes.end() ? matchGroups(groups, node_groups) : i->second;
+}
+
+std::optional<PointabilityType> Pointabilities::matchObject(const std::string &name,
+	const ItemGroupList &groups) const
+{
+	auto i = objects.find(name);
+	return i == objects.end() ? matchGroups(groups, object_groups) : i->second;
+}
+
+std::optional<PointabilityType> Pointabilities::matchGroups(const ItemGroupList &groups,
+	const std::unordered_map<std::string, PointabilityType> &pointable_groups)
+{
+	// prefers POINTABLE over POINTABLE_NOT over POINTABLE_BLOCKING
+	bool blocking = false;
+	bool not_pontable = false;
+	for (auto const &ability : pointable_groups) {
+		if (itemgroup_get(groups, ability.first) > 0) {
+			switch(ability.second) {
+				case PointabilityType::POINTABLE:
+					return PointabilityType::POINTABLE;
+					break;
+				case PointabilityType::POINTABLE_NOT:
+					not_pontable = true;
+					break;
+				default:
+					blocking = true;
+					break;
+			}
+		}
+	}
+	if (not_pontable)
+		return PointabilityType::POINTABLE_NOT;
+	if (blocking)
+		return PointabilityType::POINTABLE_BLOCKING;
+	return {};
```

Comment:
This had me confused for a sec, I'd usually use `std::nullopt` here.

Reply:
Replaced.